0015-automatic-cluster-upgrades/0015-automatic-cluster-upgrades.md

```diff
@@ -0,0 +1,139 @@
+
+Giant Swarm designed a versioning scheme (based on Semver), which includes all components on the stack, to ensure idempotence and expose the level of impact that changes bring to the customer. Patches and Minor releases should have no impact on the customer workloads, which means that they can be applied in the environments at any time.
+
+## Current state
+
+Right now the automatic upgrades are carried by Solution Engineers agreeing with the customer on upgrade windows and environments selection. The idea is to automate all this behavior and information under a Kubernetes operator(s) and Custom Resource(s).
+
+The Solution engineer plans cluster upgrades based on cluster type (environment, team,...) and maintenance windows offered by the customers. The upgrade is just a change on the cluster version field on the cluster Custom Resource (CR).
+
+Based on the fact Giant Swarm is moving to Cluster API this is story let us define a new greenfield scenario where we can design a model that can be more generic and not focus only in Giant Swarm platform. The idea would be contribute as much as possible with upstream due the fact most of this functionality would be useful for everyone that manages clusters.
+
+## Concepts
+
+
+### Cluster upgrade entity
+
+There are different reason why we are interested in having an object to define the upgrade intent. 
+
+- As we define in the user stories, we/customers would like to schedule upgrades on certain date from running version to a specific one. That way customer can plan upgrades ahead and so do we.
+
+- We have a registry of upgrades done to a cluster for auditing and tracking purposes.
+
+- We can save cluster upgrade status displaying if something goes wrong or not.
+
+- In theory we can include apps upgrades together or separately.
+
+### Upgrade policy entity
+
+Our customer can have a different set of constraints when it comes to allow changes in their clusters. The upgrade policy would define those and then the clusters can rely on a policy to enforce the desired behavior.
+
+The upgrade policy would be possible to be referenced from different cluster resources (Cluster,Machine Pool, KubeadmControlPlane, Apps,...) letting controllers to decide if the resource would be or not reconciled.
+
+The upgrade policy would contain
+
+- A way to define dates or times where clusters (or relatives) can be modified.
+
+- A way to define dates or times where clusters (or relatives) cannot be modified.
+
+## Technical aspects
+
+We envision aforementioned concepts would become custom resource definitions and there will be additional controller that would enhance and ensure the correct behavior.
+
+### Resources
+
+#### ClusterUpgrade
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: ClusterUpgrade
+metadata:
+  name: upgrade-to-release-v20
+  namespace: default
+spec:
+  clusterRef:
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: foo01
+  kubernetesVersion: 1.19.8
```

Comment:
Hmm, actually just thinking about it now, our _upgrade intent_ resource should only specify the Giant Swarm release we want to upgrade to, right? And _upgrade-operator_ will figure out the versions of components to set based on `Release` CR?

Reply:
We had a chat and decided that yes, this should only reference the `Release` that should be upgraded to, in order to:

- avoid duplicating the information about components between `Release` and `ClusterUpgrade`
- avoid introducing a way to side-step the linking between component versions encoded in a `Release` (as someone could create a `ClusterUpgrade` with versions whose combination is not encoded in a `Release` and so is untested)

I've updated my suggestion below.